agent/consul/state/catalog.go

```diff
@@ -355,9 +355,14 @@ func (s *Store) ensureNodeTxn(tx *memdb.Txn, idx uint64, node *structs.Node) err
 		if existing != nil {
 			n = existing.(*structs.Node)
 			if n.Node != node.Node {
-				return fmt.Errorf("node ID %q for node %q aliases existing node %q",
-					node.ID, node.Node, n.Node)
+				// We are actually renaming a node, remove its reference first
+				err := s.deleteNodeTxn(tx, idx, n.Node)
```

Comment:
@pierresouchay One concern we have is that this will allow multiple nodes to use the same name with different UUIDs. This might break some things and at the very least could be confusing with regards to metrics and labels. 

I think this could be remedied by refactoring this block of code and the lookup by name right after this. Essentially we would want to only delete the node when it has an id, the name is different **and** the name is not already associated with another node.

Reply:
@mkeeler Great, I made the fix, I also fix another issue regarding case insensitivity that lead us to a major downtime of Consul.

Since DNS is case insensitive and DB as issues when similar names with different cases are added, check for unicity based on case insensitivity.

Following another big incident we had in our cluster, we also validate that adding/renaming a not does not conflicts with case insensitive matches.

We had the following error once:

* one node called: mymachine.MYDC.mydomain was shut off
* another node (different ID) was added with name: mymachine.mydc.mydomain before 72 hours

When restarting the consul server of domain, the consul server restarted failed to start since it detected an issue in RAFT database because mymachine.MYDC.mydomain and mymachine.mydc.mydomain had the same names.

Checking at registration time with case insensitivity should definitly fix those issues and avoid Consul DB corruption.